### PR TITLE
[Distributed] Moving to RapidJSON

### PR DESCRIPTION
**Pull request by obelisk**
This is port of all of the distributed (OnDemand) code form boost:ptree to rapidjson.

### CODE REVIEW

**Comment by osqueryer** [comment]
:no_good: The commit 92df0383a62b86e94e2b30472ab591d6390e352d (Job results: [1505](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1505/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 92df0383a62b86e94e2b30472ab591d6390e352d (Job results: [4647](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4647/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 92df0383a62b86e94e2b30472ab591d6390e352d (Job results: [4648](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4648/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 92df0383a62b86e94e2b30472ab591d6390e352d (Job results: [1537](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1537/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 92df0383a62b86e94e2b30472ab591d6390e352d (Job results: [1538](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1538/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 86409585d46c72586e3c5ee2304181bb128ed698 (Job results: [1552](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1552/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 86409585d46c72586e3c5ee2304181bb128ed698 (Job results: [4710](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4710/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 86409585d46c72586e3c5ee2304181bb128ed698 (Job results: [4711](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4711/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 86409585d46c72586e3c5ee2304181bb128ed698 (Job results: [1591](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1591/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 86409585d46c72586e3c5ee2304181bb128ed698 (Job results: [1592](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1592/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 521274a3894001cb30afaaff197dad38549242d2 (Job results: [1553](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1553/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 521274a3894001cb30afaaff197dad38549242d2 (Job results: [4712](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4712/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 521274a3894001cb30afaaff197dad38549242d2 (Job results: [1593](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1593/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 521274a3894001cb30afaaff197dad38549242d2 (Job results: [4713](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4713/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 521274a3894001cb30afaaff197dad38549242d2 (Job results: [1594](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1594/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 8e82ad2168ccbb54f2e1dd1185dbd38c72b1ecbb (Job results: [1583](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1583/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 8e82ad2168ccbb54f2e1dd1185dbd38c72b1ecbb (Job results: [4763](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4763/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 8e82ad2168ccbb54f2e1dd1185dbd38c72b1ecbb (Job results: [4764](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4764/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 8e82ad2168ccbb54f2e1dd1185dbd38c72b1ecbb (Job results: [1634](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1634/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 8e82ad2168ccbb54f2e1dd1185dbd38c72b1ecbb (Job results: [1635](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1635/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 1e08873e587e5b083968463f9b87aaa34ff494cb (Job results: [1597](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1597/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 1e08873e587e5b083968463f9b87aaa34ff494cb (Job results: [4779](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4779/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 1e08873e587e5b083968463f9b87aaa34ff494cb (Job results: [4780](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4780/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 1e08873e587e5b083968463f9b87aaa34ff494cb (Job results: [1650](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1650/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 1e08873e587e5b083968463f9b87aaa34ff494cb (Job results: [1651](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1651/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit a0a80ffeb416a7976a30921c2c7d0cc6ec58aaac (Job results: [1630](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1630/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit a0a80ffeb416a7976a30921c2c7d0cc6ec58aaac (Job results: [1691](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1691/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit a0a80ffeb416a7976a30921c2c7d0cc6ec58aaac (Job results: [1692](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1692/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 27191bb4bc03f362a8949c39d23ad28f42272e30 (Job results: [1792](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1792/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 27191bb4bc03f362a8949c39d23ad28f42272e30 (Job results: [1884](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1884/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 27191bb4bc03f362a8949c39d23ad28f42272e30 (Job results: [1885](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1885/)) failed one or more tests (**Windows**).

**Comment by groob** [comment]
@jedi22 do you know if this change will have potential impact on the format of the serialized JSON? 

The current ad-hoc queries are all serialized as strings, even for columns which are not text. I'm just curious if a new JSON library has the potential to change that.

**Comment by jedi22** [comment]
@groob I believe this will make the output types what they should be. boost::ptree gives impossible output sometimes (see the issue on moving to RapidJSON) so hopefully this will allow us to make stronger guarantees about osquery. Is this going to cause issues for you?

**Comment by groob** [comment]
Nothing specific, but it's definitely a pull request I'm going to have to test. There's been a few times where we've noticed weird output from the current encoder and special cases it.

**Comment by groob** [comment]
Will keep an eye and report if I see something odd.

**Comment by osqueryer** [comment]
:no_good: The commit bb202406a48493685749d80d88c77622ba901a2c (Job results: [1902](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1902/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit bb202406a48493685749d80d88c77622ba901a2c (Job results: [5140](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5140/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit bb202406a48493685749d80d88c77622ba901a2c (Job results: [149](https://jenkins.osquery.io/job/osqueryPullRequestLinux/149/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit bb202406a48493685749d80d88c77622ba901a2c (Job results: [150](https://jenkins.osquery.io/job/osqueryPullRequestLinux/150/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:no_good: The commit e637f2141c83a60dbe4d5181ff92d1b2856d0c9c (Job results: [1907](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1907/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit e637f2141c83a60dbe4d5181ff92d1b2856d0c9c (Job results: [155](https://jenkins.osquery.io/job/osqueryPullRequestLinux/155/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit e637f2141c83a60dbe4d5181ff92d1b2856d0c9c (Job results: [156](https://jenkins.osquery.io/job/osqueryPullRequestLinux/156/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:no_good: The commit 9fc87cf83c3294b17d5e6de9e2346d058998fa6c (Job results: [1908](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1908/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 9fc87cf83c3294b17d5e6de9e2346d058998fa6c (Job results: [157](https://jenkins.osquery.io/job/osqueryPullRequestLinux/157/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 3b6ad40ef691eebb04969dc2dc8d3afd0fc59226 (Job results: [158](https://jenkins.osquery.io/job/osqueryPullRequestLinux/158/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 3b6ad40ef691eebb04969dc2dc8d3afd0fc59226 (Job results: [5149](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5149/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 3b6ad40ef691eebb04969dc2dc8d3afd0fc59226 (Job results: [159](https://jenkins.osquery.io/job/osqueryPullRequestLinux/159/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 3b6ad40ef691eebb04969dc2dc8d3afd0fc59226 (Job results: [5151](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5151/)) failed one or more tests (**Linux**).

**Comment by theopolis** [comment]
Looks good! I'd recommend a pass through to clean up all `rapidjson` namespace and `using`-alias them to `rj`.

**Comment by osqueryer** [comment]
:no_good: The commit d945d7cf90ecb9638f8fabeeefba072f25415a9b (Job results: [1986](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1986/)) failed the code audit or documentation test.

File: osquery/distributed/tests/distributed_tests.cpp [hunk 1]
```diff
@@ -24,6 +24,10 @@
 #include "osquery/tests/test_additional_util.h"
 #include "osquery/tests/test_util.h"
 
+#include <rapidjson/prettywriter.h>
+
+#undef GetObject
```
Comment (theopolis) [inline osquery/distributed/tests/distributed_tests.cpp]: Nit, I'd recommend documenting why this is `undef`ined.

File: osquery/distributed/distributed.cpp [hunk 2]
```diff
@@ -17,10 +17,17 @@
 #include <osquery/sql.h>
 #include <osquery/system.h>
 
+#include <rapidjson/document.h>
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]: Why aren't these defines placed into `osquery/core/json.h`?

File: osquery/distributed/distributed.cpp [hunk 3]
```diff
@@ -17,10 +17,17 @@
...
+#include <rapidjson/error/en.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+
 #include "osquery/core/conversions.h"
 #include "osquery/core/json.h"
 
-namespace pt = boost::property_tree;
+#undef GetObject
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]: Is this relevant every where RJ is used?

File: osquery/database/database.cpp [hunk 4]
```diff
@@ -724,4 +817,67 @@ void dumpDatabase() {
...
 }
+
+Status serializeQueryDataRJ(const QueryData& q, rj::Document& d) {
+  if (!d.IsArray()) {
+    return Status(1, "Document is not an array");
+  }
+  for (const auto& r : q) {
+    rj::Document serialized;
+    serialized.SetObject();
+    auto s = serializeRowRJ(r, serialized);
```
Comment (theopolis) [inline osquery/database/database.cpp]: Just an FYI, there's an inconsistent use of `s` and `status` for the local `Status` variables.

File: osquery/database/benchmarks/database_benchmarks.cpp [hunk 5]
```diff
@@ -159,4 +219,25 @@ static void DATABASE_store_append(benchmark::State& state) {
...
+  size_t k = 0;
+  while (state.KeepRunning()) {
+    setDatabaseValue(kPersistentSettings, "key" + std::to_string(k), content);
+    deleteDatabaseValue(kPersistentSettings, "key" + std::to_string(k));
+    k++;
+  }
+
+  // All benchmarks will share a single database handle.
+  for (size_t i = 0; i < k; ++i) {
+    // deleteDatabaseValue(kPersistentSettings, "key" + std::to_string(i));
```
Comment (theopolis) [inline osquery/database/benchmarks/database_benchmarks.cpp]: Why is this removed?

File: include/osquery/database.h [hunk 6]
```diff
@@ -20,6 +20,11 @@
 #include <osquery/registry.h>
 #include <osquery/status.h>
 
+#include <rapidjson/document.h>
```
Comment (theopolis) [inline include/osquery/database.h]: Nit, again, why not move includes into the json.h header?

File: osquery/distributed/distributed.cpp (hunk 3 above)
Comment (obelisk) [inline osquery/distributed/distributed.cpp]: Yes because RapidJSON defines GetObject which is defined globally on Windows causing compilation errors.

**Comment by osqueryer** [comment]
:no_good: The commit 46dcda17e413631437cf238712c4687eead13d9b (Job results: [2136](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2136/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 041a4a861717c49400f97bc17ed95f7c3c4456b7 (Job results: [2309](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2309/)) failed one or more tests (**Windows**).

**Comment by theopolis** [comment]
If you rebase, this should be ready to merge.
